ember-core/src/test/scala/org/http4s/ember/core/StreamingParserSuite.scala

```diff
@@ -0,0 +1,198 @@
+            val prefix = out.take(idx)
+            val curr = out(idx)
+            val suffix = out.drop(idx + 1)
+
+            Gen.chooseNum(0, curr.length - 1).flatMap { splitIdx =>
+              val (l, r) = curr.splitAt(splitIdx)
+              val split = List(l, r).filterNot(_.isEmpty)
+              val next = List(prefix, split, suffix).filterNot(_.isEmpty).flatten
+              go(next, remaining - 1)
+            }
+          }
+        }
+
+      if (as.isEmpty) Gen.const(Nil) else Gen.delay(go(List(as), count))
+    }
+
+    def debug(segments: List[List[Byte]]): Unit = {
+      // handleErrorWith(e => {e.printStackTrace();IO(false)}).assert
+
+      println("----")
+      segments.foreach { s =>
+        println(s)
+      }
+      // println("------")
+      // segments.foreach { s =>
+      //   s.foreach { byte =>
+      //     if (byte == 13) print("CR")
+      //     else if (byte == 10) print("NL")
+      //     else print(new String(Array[Byte](byte)))
+      //   }
+      //   println()
+      // }
+    }
+  }
+
+  object Fixtures {
+    val RequestFixedBody = toBytes(
+      List(
+        "POST /foo HTTP/1.1\r\n",
+        "Content-Type: text/plain\r\n",
+        "Content-Length: 5\r\n\r\n",
+        "hello"
+      ))
+
+    val RequestChunkedBody = toBytes(
+      List(
+        "POST /foo HTTP/1.1\r\n",
+        "Content-Type: text/plain\r\n",
+        "Transfer-Encoding: chunked\r\n\r\n",
+        "7\r\n",
+        "Mozilla\r\n",
+        "9\r\n",
+        "Developer\r\n",
+        "7\r\n",
+        "Network\r\n",
+        "0\r\n",
+        "\r\n"
+      ))
+
+    val ResponseFixedBody = toBytes(
+      List(
+        "HTTP/1.1 200 OK\r\n",
+        "Content-Type: text/plain\r\n",
+        "Content-Length: 5\r\n\r\n",
+        "hello"
+      ))
+
+    val ResponseVariableBody = toBytes(
+      List(
+        "HTTP/1.1 200 OK\r\n",
+        "Content-Type: text/plain\r\n",
+        "Transfer-Encoding: chunked\r\n\r\n",
+        "7\r\n",
+        "Mozilla\r\n",
+        "9\r\n",
+        "Developer\r\n",
+        "7\r\n",
+        "Network\r\n",
+        "0\r\n",
+        "\r\n"
+      ))
+
+    // TODO: Express in terms of Gen[Request[IO]] and Gen[Response[IO]]
+    def genRequest: Gen[List[List[Byte]]] =
+      Gen.oneOf(RequestFixedBody, RequestChunkedBody).flatMap(Helpers.subdivided(_, 25))
+    def genResponse: Gen[List[List[Byte]]] =
+      Gen.oneOf(ResponseFixedBody, ResponseVariableBody).flatMap(Helpers.subdivided(_, 25))
+
+    def toBytes(lines: List[String]): List[Byte] =
+      lines.mkString.getBytes(java.nio.charset.StandardCharsets.US_ASCII).toList
+  }
+
+  test("parse single request") {
+    PropF.forAllNoShrinkF(Fixtures.genRequest) { segments =>
+      (for {
+        read <- Helpers.taking[IO, Byte](segments)
+        result <- Parser.Request.parser(Int.MaxValue)(Array.emptyByteArray, read)
+        _ <- result._1.body.compile.drain
+        _ <- result._2
+      } yield true).assert
```

Comment:
For now, I'm not asserting anything about the response body because we have other test suites like `TraversalSpec` that do (though that one is disabled right now). The main goal here is to assert that well-formed messages are parsed successfully regardless of their chunking.

Reply:
That being said, it would be nice to ensure that we're not dropping bytes or anything in certain chunking patterns.